File: lib/py/uapi/internal/schema/ParseErrorType.py
from typing import TYPE_CHECKING, cast
from uapi.internal.schema.SchemaParseFailure import SchemaParseFailure
from uapi.internal.types.UError import UError

if TYPE_CHECKING:
    from uapi.internal.schema.ParseContext import ParseContext


def parse_error_type(error_definition_as_parsed_json: dict[str, object],
                     schema_key: str,
                     ctx: 'ParseContext') -> 'UError':
    from uapi.UApiSchemaParseError import UApiSchemaParseError
    from uapi.internal.schema.ParseUnionType import parse_union_type

    parse_failures = []

    other_keys = set(error_definition_as_parsed_json.keys())
    other_keys.discard(schema_key)
    other_keys.discard("///")

    if other_keys:
        for k in other_keys:
            loop_path = ctx.path + [k]
            parse_failures.append(SchemaParseFailure(
                ctx.document_name, cast(list[object], loop_path), "ObjectKeyDisallowed", {}))

    if parse_failures:
        raise UApiSchemaParseError(parse_failures)

    error = parse_union_type(error_definition_as_parsed_json, schema_key, [], [],
                             ctx)

    return UError(schema_key, error)

File: lib/java/src/main/java/uapi/internal/schema/ParseErrorType.java
package uapi.internal.schema;

import static uapi.internal.schema.ParseUnionType.parseUnionType;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import uapi.UApiSchemaParseError;
import uapi.internal.types.UError;
import uapi.internal.types.UUnion;

public class ParseErrorType {
    public static UError parseErrorType(
            List<Object> path,
            Map<String, Object> errorDefinitionAsParsedJson,
            String schemaKey,
            ParseContext ctx) {

        List<SchemaParseFailure> parseFailures = new ArrayList<>();

        Set<String> otherKeys = new HashSet<>(errorDefinitionAsParsedJson.keySet());
        otherKeys.remove(schemaKey);
        otherKeys.remove("///");

        if (!otherKeys.isEmpty()) {
            for (String k : otherKeys) {
                List<Object> loopPath = new ArrayList<>(path);
                loopPath.add(k);
                parseFailures
                        .add(new SchemaParseFailure(ctx.documentName, loopPath, "ObjectKeyDisallowed",
                                new HashMap<>()));
            }
        }

        if (!parseFailures.isEmpty()) {
            throw new UApiSchemaParseError(parseFailures, ctx.uApiSchemaDocumentNamesToJson);
        }

        // Assuming parseUnionType is adapted to Java and returns UError or its

        UUnion error = parseUnionType(path, errorDefinitionAsParsedJson, schemaKey,
                new ArrayList<>(),
                new ArrayList<>(), ctx);

        return new UError(schemaKey, error);
    }
}
